### DISCUSSION TITLE
Offline Scripts

### DISCUSSION THREAD

**Discussion by newzealandpaul** [tteck/Proxmox#3869] [Ideas]
The last time I checked, scripts pull code automatically from the github repo. So even if you checkout the entire repo, you will still be pulling scripts from the github repo. The cron updater, for example, pulls new code every time it runs.

Having the scripts be able to run "offline" (by that I mean not executing bash scripts pulled dynamically from the internet) would be a useful feature. This way updated scripts can be easily audited.

I am not suggesting @tteck, who does such a incredible job, would inject anything malicious. But given how widely these scripts are deployed this repo is likely a target by malicious actors.

**Comment by oOStroudyOo** [comment]
So are you suggesting that you'd like to run the script once, which will store the code locally on your device, and then you'll just continue to use that locally stored code instead and never touch the Tteck repo afterwards?

**Comment by newzealandpaul** [comment]
@oOStroudyOo I can see the confusion. I would like to be able to 

1. Checkout the  **https://github.com/tteck/Proxmox** repository
2. Run a selected script from the checked out repository using bash eg. **raw/main/ct/mongodb.sh**
3. When **raw/main/ct/mongodb.sh** is run, it runs locally and does not pull in any scripts from https://raw.githubusercontent.com/tteck/
4. If I want to update the scripts to the latest version, I run git pull origin

Right now in step 3. **https://raw.githubusercontent.com/tteck/Proxmox/main/misc/build.func** is pulled using curl even if the script is running from a checked out repository (Step 1.)

**Comment by Mellowlynx** [comment]
I think the best way to go for you is to deploy your own Forgejo LXC or Gitea LXC and make a “local” copy of the repository that way.
In there you can make changes and pull that script instead of the online one.
But you will need to change all pull URL's to your own “offline” one.

I also do get your security concern.
But, how the repository is currently set up, _build.func_ is the workhorse and will probably only **_tteck_** will update that one.
The service specific scripts should be easy enough to catch most of the malicious code, I would hope, before merging it.

**Comment by newzealandpaul** [comment]
@Mellowlynx yes setting up my own repo is an option, but ideally it could be incorporated into the helper scripts relatively easily. Basically check if the sourced script exists locally, and if not, use curl. Something like this:

```
FILE="main/misc/build.func"
URL="https://raw.githubusercontent.com/tteck/Proxmox/main/misc/build.func"

if [ -f "$FILE" ]; then
    echo "File '$FILE' exists. Sourcing it."
    source "$FILE"
else
    echo "File '$FILE' does not exist. Fetching and executing from URL."
    curl -s "$URL" | bash
fi

```
